Add case to handle insertion of `CtType`s

algomaster99 opened the pull request:
Fixes #64 

The changes handle insertion of a `CtType` into `CtPackage` and `CtCompilationUnit`.

algomaster99 commented on src/main/java/com/diffmin/patch/PatchApplication.java:
```diff
@@ -77,6 +77,28 @@ private static void performInsertion(
+                // Inserting into CtPackage
+                Set<CtType<?>> typesCopy =
+                        new HashSet<>(((CtPackage) toBeInserted.getParent()).getTypes());
+                for (CtType<?> type : typesCopy) {
+                    type.setParent(inWhichElement);
+                }
```
This could be combined with line 83 itself and I was trying to do using `stream.map` like so.

```java
new HashSet<>(((CtPackage) toBeInserted.getParent()).getTypes().stream().map(
        (type) -> {
                type.setParent(inWhichElement);
                return type;
         }
));
```

But it kept throwing `Bad return type in lambda expression: CtType<capture of ?> cannot be converted to R`. `R` in my  opinion should be `CtType<?>` only.

algomaster99 replied:
Also, this loop for setting parent was necessary otherwise it threw a `SpoonException`.
```sh
spoon.SpoonException: The default behavior has changed, a new check has been added! Don't worry, you can disable this check
with one of the following options:
 - by configuring Spoon with getEnvironment().setSelfChecks(true)
 - by removing the node from its previous parent (element.delete())
 - by cloning the node before adding it here (element.clone())
```

I figured that I needed to _detach_ the inserted type from their previous parent so I used `setParent` API. I was also trying to delete the parent using `type.getParent().delete()` but that did not work.

Although the solution was written in the above exception message, I was not able to implement what it was trying to suggest.

slarse replied:
> But it kept throwing Bad return type in lambda expression: CtType<capture of ?> cannot be converted to R. R in my opinion should be CtType<?> only.

This is probably a type inference failure, you may have to [_capture_ the wildcard](https://docs.oracle.com/javase/tutorial/java/generics/capture.html). Also, this is a bit of a misuse of the `map` method, you really shouldn't cause side effects in it. Streams are functional constructs, and functional constructs should be pure (no side effects). The `Stream.forEach` method is an exception: it's specifically meant for side effects in streams.

> I figured that I needed to detach the inserted type from their previous parent so I used setParent API. I was also trying to delete the parent using type.getParent().delete() but that did not work.

The error message says that you should try `element.delete()`, which translates to `type.delete()`. That removes `type` from its parent, making it safe to set it. What you've done is to remove the parent of type from its parent, which doesn't make sense in the context.

However, you should avoid modifying the new model as you're at this stage still extracting changes from it. See my suggestion below.

algomaster99 replied:
> What you've done is to remove the parent of type from its parent

I think I have removed the parent of type from the **type** itself. In my opinion, `type.setParent` removes the type from its parent and sets a new parent provided.

> However, you should avoid modifying the new model as you're at this stage still extracting changes from it.

Makes sense. I have one follow-up question. Would this still be a problem if all my patches were generated using `element.clone` instead of directly using the `element`?

algomaster99 commented on src/main/java/com/diffmin/patch/PatchGeneration.java:
```diff
@@ -83,8 +82,8 @@ private static CtElement delete(CtElement removedNode) {
                     public int getSrcNodeIndex(CtElement srcNode) {
                         CtElement srcNodeParent = srcNode.getParent();
-                        if (srcNodeParent.getValueByRole(srcNode.getRoleInParent())
-                                instanceof List) {
+                        if (srcNodeParent.getValueByRole(srcNode.getRoleInParent()) instanceof List
+                                || srcNode.getRoleInParent().equals(CtRole.CONTAINED_TYPE)) {
```
I added this or condition to handle insertions only in `CtCompilationUnit` as I wanted to take advantage of the lists it uses to store the types. In `CtPackage`, I directly set all types.

slarse replied:
Don't use `.equals` for enums, use `==`. It has the same effect, _and_ is null safe.

algomaster99 commented on src/main/java/com/diffmin/patch/PatchGeneration.java:
```diff
@@ -109,6 +108,12 @@ public int getSrcNodeIndex(CtElement srcNode) {
                                 return ((CtExecutable<?>) element.getParent()).getParameters();
+                            case CONTAINED_TYPE:
+                                CtCompilationUnit cu =
+                                        element.getFactory()
+                                                .CompilationUnit()
+                                                .getOrCreate((CtType<?>) element);
```
I had one suggestion here. Instead of using a factory method to get the compilation unit, should we implement an API for `CtType`s which would allow to fetch the compilation unit directly? The API could be something like this - `CtType.getCompilationUnit()`.

algomaster99 commented on src/test/resources/insert/class/NEW_Klass.java:
```diff
@@ -1,7 +1,11 @@
-class Klass {
-    class InsertedClass {
-        public void main() {
-            System.out.println("Doesn't accept any arguments");
-        }
```
I shifted this test case to `nested_class` directory because that made more sense.

slarse commented on src/main/java/com/diffmin/patch/PatchApplication.java:
```diff
@@ -77,6 +77,28 @@ private static void performInsertion(
                 break;
+            case CONTAINED_TYPE:
+                // Inserting into CtPackage
+                Set<CtType<?>> typesCopy =
+                        new HashSet<>(((CtPackage) toBeInserted.getParent()).getTypes());
+                for (CtType<?> type : typesCopy) {
+                    type.setParent(inWhichElement);
+                }
```
```suggestion
                Set<CtType<?>> typesCopy = ((CtType<?>) toBeInserted).getPackage().getTypes().stream()
                        .map(CtType::clone)
                        .collect(Collectors.toSet());
                ((CtPackage) inWhichElement).setTypes(typesCopy);
```

algomaster99 replied:
I wish, one day, I get a hang of streams and generic types in Java. 😅

slarse commented on src/main/java/com/diffmin/patch/PatchApplication.java:
```diff
@@ -77,6 +77,28 @@ private static void performInsertion(
+                ((CtPackage) inWhichElement).setTypes(typesCopy);
+
+                // Inserting into CtCompilationUnit
+                CtCompilationUnit cu =
+                        inWhichElement
+                                .getFactory()
+                                .CompilationUnit()
+                                .getOrCreate((CtPackage) inWhichElement);
```
This is not the right compilation unit. You're getting the compilation unit of the package (which is a `package-info.java` file). What you want is the compilation unit of the type.

algomaster99 replied:
Is getting the compilation unit of the type _more_ correct? The current implementation works for the test case. Maybe because the package in our case will contain only one two files - `package.info` and a `PREV` file.

algomaster99 replied:
I will have to give more thought to this because there can be a case when the previous model is an empty file so I cannot just `findFirst` type and get its compilation unit.

slarse commented:
@algomaster99 I don't know what is going on but half of my review has been posted, while half is still pending. Don't edit anything until I ping you again.

slarse commented on src/main/java/com/diffmin/patch/PatchApplication.java:
```diff
@@ -77,6 +78,28 @@ private static void performInsertion(
                 break;
+            case CONTAINED_TYPE:
+                // Inserting into CtPackage
+                Set<CtType<?>> typesCopy =
+                        ((CtType<?>) toBeInserted)
+                                .getPackage().getTypes().stream()
+                                        .map(CtType::clone)
+                                        .collect(Collectors.toSet());
+                ((CtPackage) inWhichElement).setTypes(typesCopy);
```
So, the comment I had here before was not supposed to be posted, I hadn't finished thinking. Anyway, this is wrong (as was the previous version), as it replaces _all_ of the types in the package. You only want to add the new type, not replace existing ones (as that renders everything else diffmin does completely redundant, you've already replaced the entire type!)

algomaster99 replied:
> as it replaces all of the types in the package

For our case, the package would be a single file (the PREV one), right?

> You only want to add the new type, not replace existing ones

I agree but this wasn't possible because types in `CtPackage` are stored in a `HashSet` so I don't have any control over the position of insertion which, apparently, is important. Thus, I had to set the entire types at once. This workaround is similar to the one we have implemented for insertion of `ThrownType`s. There will be redundant operations and we have to resolve that later. 

`CtCompilationUnit` stores its types in a list and in that case, we were able to insert a `CtType` at a precise location.

slarse reviewed (changes requested):
So, this is a good start but it's relatively incorrect overall. It works out in your single test case as the test suite is currently _very_ rough. It doesn't actually in any way check that the correct modifications have been made, just that the printed output is the same as the new model. I.e. you could implement diffmin to just return the new model, and tests would pass. When you think of it in terms of diff minimization, that doesn't even make sense, as diffmin is _supposed_ to change the output relative to the new model (it should minimize the diff).

This is all pretty telling that stronger tests are necessary for the actual patch application. I think it's appropriate that you start thinking about that right now. This PR is essentially entirely broken, so it's a very good candidate for designing a test case to break it. How can you verify that the patch applier has actually applied only the operations it should, and nothing more? You could consider using a change listener, to listen for changes to the AST and verify that only the expected elements are modified. This of course means you also need more input for the tests, so you need to figure out how to deal with that as well.

algomaster99 commented:
> as diffmin is supposed to change the output relative to the new model (it should minimize the diff).

`PatchApplication` is concerned with only patching the AST changes so as of now, it is doing its job well. There may be some cases where some changes are missed. In other words, all operations are not applied and I think I will only be able to find that out through the experiments I conduct over the next few weeks.

As for minimising is concerned, we have not even started that, right? `Diffmin` will be eventually be a combination of patch applier and a pretty printer. We are still working on the first part and when that is done, we will try to focus on the equally hard part which will minimise the diff.

> You could consider using a change listener, to listen for changes to the AST and verify that only the expected elements are modified. This of course means you also need more input for the tests, so you need to figure out how to deal with that as well.

That is a great idea. I will think about the implementation of a "change listener" when I run experiments and see them failing.

slarse commented:
> > as diffmin is supposed to change the output relative to the new model (it should minimize the diff).
> 
> `PatchApplication` is concerned with only patching the AST changes so as of now, it is doing its job well. There may be some cases where some changes are missed. In other words, all operations are not applied and I think I will only be able to find that out through the experiments I conduct over the next few weeks.

Exactly, it is only concerned with patching the AST. But the test suite doesn't test that. Here's how to prove that, replace `PatchApplication` with this:

```java
public class PatchApplication {
}
```
and replace `Main.patchAndGenerateModel` with this:

```java
    static CtModel patchAndGenerateModel(File prevFile, File newFile) throws FileNotFoundException {
        return SpoonUtil.buildModel(newFile);
    }
```

And all tests still pass. The actual application of a patch, is never directly tested. At best, the current test suite is a sanity check (which is better than nothing, to be sure!), but it's pretty weak, which I hope the above example shows.

> As for minimising is concerned, we have not even started that, right?

No, but the tests are of the entire app, and right now they don't actually make much sense for either diff minimization or patch application. I'm not saying focus on diff minimization, I'm saying focus on actually testing the patch application!

> > You could consider using a change listener, to listen for changes to the AST and verify that only the expected elements are modified. This of course means you also need more input for the tests, so you need to figure out how to deal with that as well.
> 
> That is a great idea. I will think about the implementation of a "change listener" when I run experiments and see them failing.

Just to be clear, I'm not sure it's a great idea. It was just a thought, be open for the possibility that it's a terrible idea :)

slarse commented:
Another quick idea about testing the patch application: you could check that certain elements stem from the expected file. For example, if you have an insertion of a type, you would expect that type (and all of its children) to come from the `NEW` file, while _everything outside of that type_ should come from the `OLD` file. Might be simpler to implement that using a change listener.

algomaster99 commented:
> I'm saying focus on actually testing the patch application!

The example makes it very clear :) I agree that we need to have a way to check if all operations are applied or not. After your example, the entire logic of `PatchApplication` looks like an almost pseudo-tested method xD.

Thanks for the ideas on it. I will try to come up with something on those line or maybe different in the next few days.

algomaster99 commented:
@slarse So I was trying work on improving the test suite and it does seem challenging. I was inspired from your suggestions and thought of one idea but that did not work as expected.

1. I modified the test suite so it also has access to the `Diff` object. I can further extract `rootOperations` from it and use them.
2. Next, I decided to iterate over the `rootOperations` and for each `Action` (tried for `Delete` first) I get their `srcNode` and `dstNode`.
3. For `Delete`, I wanted to check if the `srcNode` comes from the old file however I was not able to get its `CtPath` (a SpoonException was thrown). Further debugging revealed that `srcNode` gets rid of `roleInParent` (a property used to determine `CtPath`) upon its deletion (we perform deletion in `PatchApplication`). But I was still able to access parent of `srcNode` and its parent and so on.

I tried a similar approach with `UpdateOperation`s too and got the same error which led me to conclude that `rootOperation`s is probably not the way to go because we have already modified the nodes in each operation we applied.

> Might be simpler to implement that using a change listener.

I had had not heard of a change listener in context of Java before so I was a bit skeptical to look up more on that lines. Upon searching online, I found [this](https://docs.oracle.com/javase/tutorial/uiswing/events/changelistener.html). The concept sounds convincing. I will try using this to improve the test suite.

slarse commented:
>     3. For `Delete`, I wanted to check if the `srcNode` comes from the old file however I was not able to get its `CtPath` (a SpoonException was thrown). Further debugging revealed that `srcNode` gets rid of `roleInParent` (a property used to determine `CtPath`) upon its deletion (we perform deletion in `PatchApplication`). But I was still able to access parent of `srcNode` and its parent and so on.

`getRoleInParent()` uses the parent to find the role of the element. When you call `delete()` on an element, it's removed from its parent, and thus `getRoleInParent()` no longer works. Like I've mentioned before, you need to use 

> I tried a similar approach with `UpdateOperation`s too and got the same error which led me to conclude that `rootOperation`s is probably not the way to go because we have already modified the nodes in each operation we applied.

I'd suggest that you write down the CtPaths to elements that you expect to come from the new model in a metadata file in each test directory, and then just iterate over the entire tree with a scanner. Every time you come across an element that matches one of the stored CtPaths, you assert that it comes from the new model. For other elements, you assert that they come from the old model. You'll probably need to tweak the approach a bit to not be too strict.

> > Might be simpler to implement that using a change listener.
> 
> I had had not heard of a change listener in context of Java before so I was a bit skeptical to look up more on that lines. Upon searching online, I found [this](https://docs.oracle.com/javase/tutorial/uiswing/events/changelistener.html). The concept sounds convincing. I will try using this to improve the test suite.

Note Spoon has built-in support for change listeners. Also note that the change will almost always be registered on the parent of the element you're interested in. I think using a less dynamic approach (the one I suggest above) is easier.

algomaster99 commented:
> When you call delete() on an element, it's removed from its parent

Yes, I am aware of that but if that is the case why was `element.getParent()` still working? In fact, `element.isParentInitiazed()` was still returning `true`.

> Note Spoon has built-in support for change listeners

Could you share some code snippets which implement so?

algomaster99 commented:
```java
File f1 = sources.prevPath.toFile();
File f2 = sources.newPath.toFile();

List<String> lines = Files.readAllLines(sources.modifiedPaths); // All paths that are from the new model line-by-line

CtModel patchedCtModel = Main.patchAndGenerateModel(f1, f2);

CtPackage prevPackage = SpoonUtil.buildModel(f1).getRootPackage();
CtPackage newPackage = SpoonUtil.buildModel(f2).getRootPackage();

Iterator<CtElement> it = patchedCtModel.getRootPackage().descendantIterator();

while (it.hasNext()) {
    CtElement element = it.next();
    if (element instanceof CtPackage) { // handle SpoonException when getPath is called on an instance of CtPackage
        continue;
     }
    CtPath path = element.getPath();
    if (!lines.stream().filter(line -> line.equals(path.toString())).findFirst().isEmpty()) { // check if any of the path in metadata file match in patched program
        assertTrue(path.evaluateOn(prevPackage).isEmpty()); // the matched path should not be there in prev model
        assertTrue(path.evaluateOn(newPackage).size() == 1); // the matched should be there in new model and should match only once
    }
}
```

This is how I have tried to modify the `AppTest` and it works for `insert/literal`. I will extend this approach for others when we have some further discussion on it.

I think this approach will work for insert and update operations but I am not sure how we can test for delete patches as the nodes corresponding to it don't even exist in the patched file. One way which I was pondering over was to store the path of deleted paths in another metadata file and just check that these paths are no longer there in the patched file.

algomaster99 commented:
@slarse I had one question regarding container of `CtType` in `CtPackage`. The container is a `Set` and I wanted to ask why making it a set was necessary. Logically, two identical classes can exist in a packages which are nested as their names can be resolved via qualified names of packages.

In fact, `CtCompilationUnit`'s type's container should have been a `Set` (it is a `List` as of now) because all types in one file should be unique.

The set container of `CtType` in `CtPackage` is also posing a problem. Since it's a set, we cannot insert a type at particular index so we have to replace all types using [`setTypes`](https://github.com/INRIA/spoon/blob/93e82b0823e3eba7ffdbcce0d87871fb7dffb5ad/src/main/java/spoon/support/reflect/declaration/CtPackageImpl.java#L178). This will fail the element origin test but I can workaround that like I have done for thrown types.

slarse commented:
> Logically, two identical classes can exist in a packages which are nested as their names can be resolved via qualified names of packages.

Packages cannot have duplicated class names in Java. Do you mean e.g. that there can be `diffmin.Main` and `diffmin.util.Main`? `diffmin` and `diffmin.util` are _different_ packages.

> In fact, `CtCompilationUnit`'s type's container should have been a `Set` (it is a `List` as of now) because all types in one file should be unique.

It is a list because in a compilation unit, the types have order (the order they are declared in). In terms of patch application, we should be very glad it's a list because it allows us to preserve the order in which types are defined within a file.

> The set container of `CtType` in `CtPackage` is also posing a problem. Since it's a set, we cannot insert a type at particular index so we have to replace all types using [`setTypes`](https://github.com/INRIA/spoon/blob/93e82b0823e3eba7ffdbcce0d87871fb7dffb5ad/src/main/java/spoon/support/reflect/declaration/CtPackageImpl.java#L178). This will fail the element origin test but I can workaround that like I have done for thrown types.

There's no problem with order in a `CtPackage` as that's not reflected anywhere when printing. The order of types in a _file_ is the order in the compilation unit's list. The types of a package typically exist in many different files, and cross-file ordering doesn't make much sense.

Is there not a `CtPackage.addType()` method?

algomaster99 commented:
> Do you mean e.g. that there can be diffmin.Main and diffmin.util.Main?

Yes. But `diffmin.util` is a sub-package of `diffmin`, right?

> we should be very glad it's a list because it allows us to preserve the order in which types are defined within a file.

I agree. I just wrote this to draw a comparison between containers of types in `CtCompilationUnit` and `CtPackage`.

> Is there not a CtPackage.addType() method?

There is. I just did not use it because I felt it would just append it but we need to control the position too. If `CtCompilationUnit` takes care of that, then we can use it.

slarse commented:
> > Do you mean e.g. that there can be diffmin.Main and diffmin.util.Main?
> 
> Yes. But `diffmin.util` is a sub-package of `diffmin`, right?

Yes, but they are still different packages. The types of `diffmin.util` do not belong to `diffmin`, they belong to `diffmin.util`. For example, a class in `diffmin.util` cannot access package-private stuff from `diffmin`, and similarly, `diffmin` can't access package-private stuff from `diffmin.util`.

> > Is there not a CtPackage.addType() method?
> 
> There is. I just did not use it because I felt it would just append it but we need to control the position too. If `CtCompilationUnit` takes care of that, then we can use it.

Well, the compilation unit is pretty much the Java representation of a file, while a package is the Java representation of a directory. The order if files within a directory is defined by the file system, and so we don't need to be concerned with it.

algomaster99 commented:
If we called `getTypes` on `diffmin`, we should get all the types (`diffmin.Main` and `diffmin.util.Main`) in `diffmin` and the sub-packages, right? So if the type container is a set, won't it exclude one of those types? It doesn't happen but why not is my question.

Hold the review. The test failed and I still have to think about [this](https://github.com/SpoonLabs/diffmin/pull/65#discussion_r637055486).

algomaster99 commented:
> This is not the right compilation unit. You're getting the compilation unit of the package (which is a package-info.java file). What you want is the compilation unit of the type.

I agree with this. But according to my code, I am inserting the types in the `package-info.java` file and it still passes the tests somehow. What could be the reason for this? I know it's wrong an even JaCoCo indicates this [here](https://github.com/SpoonLabs/diffmin/pull/65/files#diff-84bbf6aff3327a303cfd11a421a820acb9423869483230a421a72f07c0e6d596R94). That else branch has to be covered as the types were never empty.

slarse commented:
> If we called `getTypes` on `diffmin`, we should get all the types (`diffmin.Main` and `diffmin.util.Main`) in `diffmin` and the sub-packages, right? So if the type container is a set, won't it exclude one of those types? It doesn't happen but why not is my question.

No, you should only get the types from `diffmin`.

> What could be the reason for this?

The test suite only gets the first type from each compilation unit. It probably just happens to work out.

slarse commented:
@algomaster99 Don't force push!

algomaster99 commented:
> Don't force push!

I had to. The force-push was just for rebasing the new changes in our test suite.

slarse commented:
> > Don't force push!
> 
> I had to. The force-push was just for rebasing the new changes in our test suite.

You don't need to rebase, you can just merge.

algomaster99 commented:
> You don't need to rebase, you can just merge.

I don't like merge commits as linear history is more easy for me to follow. But anyway, this PR is going to be squashed so doesn't matter. I thought I could make an exception for force-pushing where I just want to update the PR with latest changes. 😅

slarse commented:
> > You don't need to rebase, you can just merge.
> 
> I don't like merge commits as linear history is easier for me to follow. But anyway, this PR is going to be squashed so doesn't matter. I thought I could make an exception for force-pushing where I just want to update the PR with latest changes. sweat_smile

I've no problems with force pushing _before_ reviews on a PR. After reviews have begun, force pushing muddles the commit history. Additionally, if the reviewer has fetched the branch under review (which I had), they need to delete the local branch and re-fetch it to ensure that they actually have the same branch.

Unless project guidelines say otherwise, I strongly advice you not to force-push to a PR where reviews have started. Or even to force push to an open PR.

algomaster99 commented:
> Additionally, if the reviewer has fetched the branch under review (which I had), they need to delete the local branch and re-fetch it to ensure that they actually have the same branch.

I used `git pull origin <branch name> --rebase` to resolve the changes. The rebase flag would take care of the inconsistency between local and remote branch.

> I strongly advice you not to force-push to a PR where reviews have started.

Yes. I am heeding to that. From now, I would not rebase even if I need to update the changes. :)

slarse commented:
> > Additionally, if the reviewer has fetched the branch under review (which I had), they need to delete the local branch and re-fetch it to ensure that they actually have the same branch.
> 
> When I faced such situations, I used `git pull origin <branch name> --rebase` to resolve the changes. The rebase flag would take care of the inconsistency between local and remote branch.

As a rebase can have conflicts, I'm not confident Git is guaranteed to be able to fix a force update. It appears as if Git is clever enough to figure out that there's been a force update, and it will simply accept it, but I'm not sure how that part of Git works.

algomaster99 commented:
For you, it will only have conflicts if you have locally committed something in the same branch which has been force pushed. This is a strong point supporting that I should not _compel_ you to rebase. :)

But if you don't have any commits, or changes in your working directory, your branch would be exactly like the remote branch _before_ the force push. Now when I force push after rebasing, I have already resolved the conflicts for everyone provided no one has  commits locally.

slarse commented:
> Now when I force push after rebasing, I have already resolved the conflicts (during rebase) for everyone provided no one has commits locally.

This intrigued me, because that's not how rebasing works. Git doesn't store a record of rebases anywhere other than in the reflog (which is local to each repository and does not get pushed), and it definitely doesn't store data on how to solve conflicts. It just rewrites the version history and that's that.

> But if you don't have any commits, or changes in your working directory, your branch would be exactly like the remote branch _before_ the force push. 

However, this seems to hold. Even when I deliberately set up a scenario in which a rebase _would definitely_ cause a conflict by force-pushing something completely unrelated to the remote, pulling with rebase still worked. Apparently, pulling with rebase does not necessarily rebase. If you're tracking a remote and that remote has updates, pulling will just set the state of the branch to the remote's state. So, pulling with rebase is almost as forceful as pulling with `--force`, except that the latter doesn't respect local changes while a rebase-pull does. I had just assumed that `git pull --rebase` actually ran `git-rebase`, so it was news to me that's not the case.

This is in contrast to fetching and then rebasing on the remote branch, which causes conflicts as expected.

algomaster99 commented:
> Git doesn't store a record of rebases anywhere other than in the reflog (which is local to each repository and does not get pushed), and it definitely doesn't store data on how to solve conflicts.

I did not mean that Git stores a record of rebases. By "I have already resolved the conflicts (during rebase) for everyone provided no one has commits locally.", I meant that you cannot get a conflict because you haven't done any work on the local repository. Only the history is rewritten and if during that rewriting any merge conflict appeared, I resolved that and it is contained within the commits only.

> pulling will just set the state of the branch to the remote's state (except if you have local changes, then additional stuff happens)

This is exactly what I wanted to convey in my previous comment. And yes, if you can local changes, there is a possibility of merge conflicts too.

> force-pushing something completely unrelated to the remote

Could you give an example? I am unable to imagine a scenario where you push _unrelated_ things. By unrelated change, do you mean that you push changes which are not affecting any of the files in the current PR?

>  forceful as pulling with --force

git doesn't have force pull. But I think an alternative would be:

If you only have local changes in the working directory only,
```bash
git reset HEAD --hard
git pull origin <branch>
```

If you have commits as well,
```bash
git reset origin/<branch> --hard
git pull origin <branch>
```

> I had just assumed that git pull --rebase actually ran git-rebase

You are right about this. `git rebase` is executed locally. You can even check the git documentation [here](https://git-scm.com/docs/git-pull#_description).
1. `git pull` = `git fetch` + `git merge`
2. `git pull --rebase ` = `git fetch` + `git rebase`

> This is in contrast to fetching and then rebasing on the remote branch, which causes conflicts as expected.

This is equivalent to `git pull --rebase`.

slarse commented:
> > Git doesn't store a record of rebases anywhere other than in the reflog (which is local to each repository and does not get pushed), and it definitely doesn't store data on how to solve conflicts.
> 
> I did not mean that Git stores a record of rebases. By "I have already resolved the conflicts (during rebase) for everyone provided no one has commits locally.", I meant that you cannot get a conflict because you haven't done any work on the local repository. Only the history is rewritten and if during that rewriting any merge conflict appeared, I resolved that and it is contained within the commits only.

If Git actually performed a rebase, there would still be conflicts, because you can't "resolve conflicts for others" with a rebase as it just rewrites history. If you do another _actual_ rebase on that, there will almost certainly be conflicts again. Now, when pulling with rebase, Git doesn't actually do a rebase for remote-tracking changes (see below).

> > pulling will just set the state of the branch to the remote's state (except if you have local changes, then additional stuff happens)
> 
> This is exactly what I wanted to convey in my previous comment. And yes, if you can local changes, there is a possibility of merge conflicts too.

But this is not the same as running `git-rebase`. It's a forced update.

> > force-pushing something completely unrelated to the remote
> 
> Could you give an example? I am unable to imagine a scenario where you push _unrelated_ things. By unrelated change, do you mean that you push changes which are not affecting any of the files in the current PR?

I mean that I push a completely different branch to the remote branch, with completely disconnected commit history.

> > forceful as pulling with --force
> 
> git doesn't have force pull.

Yes it does. From `man git-pull`:

```
       -f, --force
           When git fetch is used with <src>:<dst> refspec it may refuse to update the local branch as discussed in the <refspec> part of the git-fetch(1) documentation. This option overrides that
           check.
```

> > This is in contrast to fetching and then rebasing on the remote branch, which causes conflicts as expected.
> 
> This is equivalent to `git pull --rebase`.

No, it is not. I can show it to you during the next meeting, but the manual says so as well: "If there is a remote-tracking branch corresponding to the upstream branch and the upstream branch was rebased since last fetched, the rebase uses that information to avoid rebasing non-local changes." In other words, it just accepts upstream changes.

Therefore, you can get conflicts when doing `git fetch <remote> <branch> && git rebase <remote>/<branch>` that do not occur when doing `git pull --rebase`.

slarse commented:
Let's return this PR to its actual topic and discuss the finer details of Git somewhere else :)

algomaster99 commented:
> In other words, it just accepts upstream changes, and so it's impossible to get a conflict when pulling with rebase without local changes

> you can get conflicts when doing git fetch <remote> <branch> && git rebase <remote>/<branch>

> with a rebase as it just rewrites history

I will try to experiment with these cases once and see for myself before discussing further. I have limited knowledge of `git-rebase` and the documentation is not very developer-friendly 😅 

I will move further with this PR once I fix https://github.com/SpoonLabs/diffmin/issues/72.

algomaster99 commented:
@slarse Tests are failing because of the test case which inserts a type into an empty file. Apparently, spoon does not create a compilation unit for empty files.

I tried adding an inline-comment at the top of the `PREV` file but still spoon won't build its compilation unit.

algomaster99 commented on src/main/java/com/diffmin/patch/PatchApplication.java:
```diff
@@ -77,6 +78,29 @@ private static void performInsertion(
+                ((CtPackage) inWhichElement).addType((CtType<?>) toBeInserted.clone());
+
+                // Inserting into CtCompilationUnit
+                List<CtCompilationUnit> compilationUnits =
+                        List.copyOf(
+                                ((CtPackage) inWhichElement)
+                                        .getDeclaringModule()
+                                        .getFactory()
+                                        .CompilationUnit()
+                                        .getMap()
+                                        .values());
```
I plan to put this logic in `SpoonUtil` so that it can be reused.

slarse replied:
Yes, this should be a method along the lines of `getAllCompilationUnits(CtElement element)`. The implementation can then look like `element.getParent(CtModule.class).getFactory().CompilationUnit().getMap().values()`.

It could be further abstracted into `getUnambiguousCompilationUnit(CtElement)`, where it fetches only one compilation unit, and raises if there's anything but one.

slarse commented:
> @slarse Tests are failing because of the test case which inserts a type into an empty file. Apparently, spoon does not create a compilation unit for empty files.

I'm not sure if it's correct or not by Spoon to not create a compilation unit. But until we figure that out, you can just do some special handling if either model contains no Java source code. I even think we can ignore this case for now, because if there's no Java source code in one of the revisions, then there's no possibility to minimize the diff by AST manipulation. If the PREV revision is empty, then the NEW revision is the minimized diff. If the NEW revision is empty, then the NEW revision is _also_ the minimized diff (probably the empty string, but could be just that it lacks any syntactical elements).

algomaster99 commented:
> I even think we can ignore this case for now, because if there's no Java source code in one of the revisions, then there's no possibility to minimize the diff by AST manipulation.

That's a good point

slarse reviewed (approved):
lgtm

algomaster99 commented:
@slarse 
> Therefore, you can get conflicts when doing git fetch <remote> <branch> && git rebase <remote>/<branch> that do not occur when doing git pull --rebase.

I tried this myself, and you are right. `git fetch` + `git rebase` did result in a conflict; meanwhile `git pull --rebase` just accepted the upstream changes. So yeah, it is impossible to get conflicts without local _commits_. Thanks for pointing it out :)
